File: example/cpp/src/14_verify_identification.cpp
#include "example_common.hpp"
#include "rebotarm/identification.hpp"

#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

struct Dataset {
  Eigen::MatrixXd q;
  Eigen::MatrixXd dq;
  Eigen::MatrixXd ddq;
  Eigen::MatrixXd tau;
};

std::vector<std::string> split_csv(const std::string& line) {
  std::vector<std::string> out;
  std::stringstream ss(line);
  std::string cell;
  while (std::getline(ss, cell, ',')) {
    const auto start = cell.find_first_not_of(" \t\r\n");
    const auto end = cell.find_last_not_of(" \t\r\n");
    out.push_back(start == std::string::npos ? "" : cell.substr(start, end - start + 1));
  }
  return out;
}

int find_col(const std::vector<std::string>& header, const std::string& name) {
  for (int i = 0; i < static_cast<int>(header.size()); ++i) {
    if (header[i] == name) return i;
  }
  return -1;
}

int find_joint_col(const std::vector<std::string>& header, const std::string& prefix, int joint) {
  const std::vector<std::string> candidates = {
      prefix + std::to_string(joint),
      prefix + "_" + std::to_string(joint),
      prefix + ".joint_" + std::to_string(joint),
  };
  for (const auto& name : candidates) {
    const int col = find_col(header, name);
    if (col >= 0) return col;
  }
  return -1;
}

Dataset load_csv(const std::string& path, int dof) {
  std::ifstream in(path);
  if (!in) throw std::runtime_error("failed to open " + path);
  std::string line;
  if (!std::getline(in, line)) throw std::runtime_error("empty CSV");
  const auto header = split_csv(line);
  std::vector<int> q_cols, dq_cols, ddq_cols, tau_cols;
  for (int i = 1; i <= dof; ++i) {
    q_cols.push_back(find_joint_col(header, "q", i));
    dq_cols.push_back(find_joint_col(header, "dq", i));
    ddq_cols.push_back(find_joint_col(header, "ddq", i));
    tau_cols.push_back(find_joint_col(header, "tau", i));
  }
  for (const auto& cols : {q_cols, dq_cols, ddq_cols, tau_cols}) {
    for (int col : cols) {
      if (col < 0) throw std::runtime_error("CSV missing required q/dq/ddq/tau columns");
    }
  }
  std::vector<std::vector<double>> rows;
  while (std::getline(in, line)) {
    if (line.empty()) continue;
    const auto cells = split_csv(line);
    std::vector<double> row;
    for (const auto& cell : cells) row.push_back(std::stod(cell));
    rows.push_back(row);
  }
  Dataset ds;
  ds.q.resize(rows.size(), dof);
  ds.dq.resize(rows.size(), dof);
  ds.ddq.resize(rows.size(), dof);
  ds.tau.resize(rows.size(), dof);
  for (int r = 0; r < static_cast<int>(rows.size()); ++r) {
    for (int j = 0; j < dof; ++j) {
      ds.q(r, j) = rows[r][q_cols[j]];
      ds.dq(r, j) = rows[r][dq_cols[j]];
      ds.ddq(r, j) = rows[r][ddq_cols[j]];
      ds.tau(r, j) = rows[r][tau_cols[j]];
    }
  }
  return ds;
}

std::vector<double> load_beta_txt(const std::string& path) {
  std::ifstream in(path);
  if (!in) throw std::runtime_error("failed to open " + path);
  std::vector<double> values;
  double value = 0.0;
  while (in >> value) values.push_back(value);
  return values;
}

std::string read_text_file(const std::string& path) {
  std::ifstream in(path);
  if (!in) throw std::runtime_error("failed to open " + path);
  std::ostringstream buffer;
  buffer << in.rdbuf();
  return buffer.str();
}

std::string trim(const std::string& text) {
  const auto start = text.find_first_not_of(" \t\r\n");
  if (start == std::string::npos) return "";
  const auto end = text.find_last_not_of(" \t\r\n");
  return text.substr(start, end - start + 1);
}

std::string unquote(std::string text) {
  text = trim(text);
  if (text.size() >= 2 && ((text.front() == '"' && text.back() == '"') ||
                           (text.front() == '\'' && text.back() == '\''))) {
    return text.substr(1, text.size() - 2);
  }
  return text;
}

std::optional<std::string> yaml_value_tail(const std::string& yaml, const std::string& key) {
  std::istringstream in(yaml);
  std::string line;
  const std::string prefix = key + ":";
  while (std::getline(in, line)) {
    const std::string stripped = trim(line);
    if (stripped.rfind(prefix, 0) == 0) return trim(stripped.substr(prefix.size()));
  }
  return std::nullopt;
}

std::string yaml_string(const std::string& yaml, const std::string& key,
                        const std::string& default_value = "") {
  const auto value = yaml_value_tail(yaml, key);
  return value ? unquote(*value) : default_value;
}

bool yaml_bool(const std::string& yaml, const std::string& key, bool default_value) {
  const auto value = yaml_value_tail(yaml, key);
  if (!value) return default_value;
  const std::string lower = [&] {
    std::string out = trim(*value);
    std::transform(out.begin(), out.end(), out.begin(),
                   [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
    return out;
  }();
  return lower == "true" || lower == "1" || lower == "yes" || lower == "on";
}

std::vector<double> parse_number_list_body(std::string body) {
  for (char& ch : body) {
    if (ch == ',' || ch == '[' || ch == ']') ch = ' ';
  }
  std::istringstream in(body);
  std::vector<double> values;
  double value = 0.0;
  while (in >> value) values.push_back(value);
  return values;
}

std::vector<double> yaml_float_vector(const std::string& yaml, const std::string& key) {
  std::istringstream in(yaml);
  std::string line;
  const std::string prefix = key + ":";
  bool in_block = false;
  std::string body;
  while (std::getline(in, line)) {
    const std::string stripped = trim(line);
    if (!in_block && stripped.rfind(prefix, 0) == 0) {
      std::string tail = trim(stripped.substr(prefix.size()));
      if (tail.find('[') != std::string::npos) {
        body += tail;
        while (body.find(']') == std::string::npos && std::getline(in, line)) body += " " + trim(line);
        return parse_number_list_body(body);
      }
      in_block = true;
      continue;
    }
    if (in_block) {
      if (stripped.rfind("- ", 0) != 0) break;
      body += " " + stripped.substr(2);
    }
  }
  return parse_number_list_body(body);
}

std::vector<int> yaml_int_vector(const std::string& yaml, const std::string& key) {
  const auto floats = yaml_float_vector(yaml, key);
  std::vector<int> values;
  values.reserve(floats.size());
  for (double value : floats) values.push_back(static_cast<int>(value));
  return values;
}

Eigen::VectorXd vector_to_eigen(const std::vector<double>& values) {
  Eigen::VectorXd out(values.size());
  for (int i = 0; i < out.size(); ++i) out[i] = values[i];
  return out;
}

}  // namespace

int main(int argc, char** argv) {
  try {
    if (example::has_flag(argc, argv, "--help") || example::has_flag(argc, argv, "-h")) {
      std::cout << "Usage: ./14_verify_identification --data id_verify.csv "
                   "--params identified.yaml [--urdf robot.urdf]\n"
                   "   or: ./14_verify_identification --data id_verify.csv "
                   "--beta beta.txt [--urdf robot.urdf] [--no-friction]\n";
      return 0;
    }
    const std::string data_path = example::arg_value(argc, argv, "--data");
    const std::string params_path = example::arg_value(argc, argv, "--params");
    const std::string beta_path = example::arg_value(argc, argv, "--beta");
    if (data_path.empty() || (params_path.empty() && beta_path.empty())) {
      throw std::runtime_error("--data and either --params or --beta are required");
    }

    std::string yaml;
    std::string mode = "full";
    bool include_friction = !example::has_flag(argc, argv, "--no-friction");
    double coulomb_eps = example::arg_double(argc, argv, "--coulomb-eps", 1e-3);
    std::vector<double> beta_values;
    std::vector<int> selected_columns;
    std::string urdf_path = example::arg_value(argc, argv, "--urdf");

    if (!params_path.empty()) {
      yaml = read_text_file(params_path);
      mode = yaml_string(yaml, "mode", "full");
      include_friction = yaml_bool(yaml, "include_friction", true);
      if (example::has_flag(argc, argv, "--no-friction")) include_friction = false;
      const std::string coulomb = yaml_string(yaml, "coulomb_eps");
      if (!coulomb.empty()) coulomb_eps = std::stod(coulomb);
      if (urdf_path.empty()) urdf_path = yaml_string(yaml, "input_urdf", example::default_urdf_path());
      if (mode == "base") {
        beta_values = yaml_float_vector(yaml, "beta_base");
        if (beta_values.empty()) beta_values = yaml_float_vector(yaml, "beta");
        selected_columns = yaml_int_vector(yaml, "selected_columns");
      } else if (mode == "full") {
        beta_values = yaml_float_vector(yaml, "beta");
      } else {
        throw std::runtime_error("--params must contain mode full or base");
      }
    } else {
      beta_values = load_beta_txt(beta_path);
      if (urdf_path.empty()) urdf_path = example::urdf_arg(argc, argv);
    }
    if (beta_values.empty()) throw std::runtime_error("no beta values found");

    rebotarm::RobotModel model(urdf_path);
    Dataset ds = load_csv(data_path, model.nv());
    const Eigen::MatrixXd Y = rebotarm::ident::build_regression_matrix(
        model, ds.q, ds.dq, ds.ddq, include_friction, coulomb_eps);
    const Eigen::VectorXd beta = vector_to_eigen(beta_values);

    const Eigen::VectorXd tau = rebotarm::ident::stack_tau_samples(ds.tau);
    Eigen::VectorXd tau_pred;
    if (mode == "base") {
      if (selected_columns.empty()) throw std::runtime_error("base params require selected_columns");
      if (beta.size() != static_cast<int>(selected_columns.size())) {
        throw std::runtime_error("base beta length does not match selected_columns");
      }
      Eigen::MatrixXd Y_selected(Y.rows(), selected_columns.size());
      for (int i = 0; i < static_cast<int>(selected_columns.size()); ++i) {
        if (selected_columns[i] < 0 || selected_columns[i] >= Y.cols()) {
          throw std::runtime_error("selected column out of range");
        }
        Y_selected.col(i) = Y.col(selected_columns[i]);
      }
      tau_pred = Y_selected * beta;
    } else {
      if (beta.size() != Y.cols()) throw std::runtime_error("beta length does not match regressor columns");
      tau_pred = Y * beta;
    }
    const auto metrics = rebotarm::ident::regression_metrics(tau, tau_pred, model.nv());
    std::cout << "verify samples=" << ds.q.rows() << " rmse=" << metrics.rmse
              << " mae=" << metrics.mae << " max_abs=" << metrics.max_abs
              << " r2=" << metrics.r2 << "\n";
    std::cout << "per-joint rmse: " << metrics.per_joint_rmse.transpose() << "\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "error: " << e.what() << "\n";
    return 1;
  }
}

File: example/rust/src/bin/14_verify_identification.rs
use rebotarm_control_rt_rust_examples::common::{
    arg_value, has_flag, parse_float_arg, parse_urdf_path, MathModel,
};
use std::env;
use std::error::Error;
use std::fs;
use std::path::PathBuf;

struct Dataset {
    q: Vec<f64>,
    dq: Vec<f64>,
    ddq: Vec<f64>,
    tau: Vec<f64>,
    samples: usize,
}

fn find_col(header: &[&str], prefix: &str, joint: usize) -> Result<usize, Box<dyn Error>> {
    let names = [
        format!("{prefix}{joint}"),
        format!("{prefix}_{joint}"),
        format!("{prefix}.joint_{joint}"),
    ];
    header
        .iter()
        .position(|item| names.iter().any(|name| name == item))
        .ok_or_else(|| format!("missing CSV column for {prefix}{joint}").into())
}

fn load_csv(path: &str, dof: usize) -> Result<Dataset, Box<dyn Error>> {
    let text = fs::read_to_string(path)?;
    let mut lines = text.lines();
    let header_line = lines.next().ok_or("empty CSV")?;
    let header: Vec<&str> = header_line.split(',').map(str::trim).collect();
    let q_cols: Vec<_> = (1..=dof)
        .map(|i| find_col(&header, "q", i))
        .collect::<Result<_, _>>()?;
    let dq_cols: Vec<_> = (1..=dof)
        .map(|i| find_col(&header, "dq", i))
        .collect::<Result<_, _>>()?;
    let ddq_cols: Vec<_> = (1..=dof)
        .map(|i| find_col(&header, "ddq", i))
        .collect::<Result<_, _>>()?;
    let tau_cols: Vec<_> = (1..=dof)
        .map(|i| find_col(&header, "tau", i))
        .collect::<Result<_, _>>()?;

    let mut q = Vec::new();
    let mut dq = Vec::new();
    let mut ddq = Vec::new();
    let mut tau = Vec::new();
    let mut samples = 0;
    for line in lines {
        if line.trim().is_empty() {
            continue;
        }
        let values: Vec<f64> = line
            .split(',')
            .map(|cell| cell.trim().parse::<f64>())
            .collect::<Result<_, _>>()?;
        for idx in &q_cols {
            q.push(values[*idx]);
        }
        for idx in &dq_cols {
            dq.push(values[*idx]);
        }
        for idx in &ddq_cols {
            ddq.push(values[*idx]);
        }
        for idx in &tau_cols {
            tau.push(values[*idx]);
        }
        samples += 1;
    }
    Ok(Dataset {
        q,
        dq,
        ddq,
        tau,
        samples,
    })
}

fn trim(text: &str) -> &str {
    text.trim_matches(|ch: char| ch.is_whitespace())
}

fn unquote(text: &str) -> String {
    let text = trim(text);
    if text.len() >= 2
        && ((text.starts_with('"') && text.ends_with('"'))
            || (text.starts_with('\'') && text.ends_with('\'')))
    {
        text[1..text.len() - 1].to_string()
    } else {
        text.to_string()
    }
}

fn yaml_value<'a>(yaml: &'a str, key: &str) -> Option<&'a str> {
    let prefix = format!("{key}:");
    yaml.lines().find_map(|line| {
        let stripped = trim(line);
        stripped.strip_prefix(&prefix).map(|value| trim(value))
    })
}

fn yaml_string(yaml: &str, key: &str, default: &str) -> String {
    yaml_value(yaml, key)
        .map(unquote)
        .unwrap_or_else(|| default.to_string())
}

fn yaml_bool(yaml: &str, key: &str, default: bool) -> bool {
    yaml_value(yaml, key)
        .map(|value| {
            matches!(
                trim(value).to_ascii_lowercase().as_str(),
                "true" | "1" | "yes" | "on"
            )
        })
        .unwrap_or(default)
}

fn parse_number_list_body(body: &str) -> Vec<f64> {
    body.replace([',', '[', ']'], " ")
        .split_whitespace()
        .filter_map(|value| value.parse::<f64>().ok())
        .collect()
}

fn yaml_float_vector(yaml: &str, key: &str) -> Vec<f64> {
    let prefix = format!("{key}:");
    let mut in_block = false;
    let mut body = String::new();
    for line in yaml.lines() {
        let stripped = trim(line);
        if !in_block {
            if let Some(tail) = stripped.strip_prefix(&prefix) {
                let tail = trim(tail);
                if tail.contains('[') {
                    body.push_str(tail);
                    if tail.contains(']') {
                        break;
                    }
                    in_block = true;
                } else {
                    in_block = true;
                }
            }
            continue;
        }
        if body.contains(']') {
            break;
        }
        if let Some(item) = stripped.strip_prefix("- ") {
            body.push(' ');
            body.push_str(item);
        } else if stripped.contains(']') || stripped.contains('[') || stripped.contains(',') {
            body.push(' ');
            body.push_str(stripped);
        } else if !body.is_empty() {
            break;
        }
    }
    parse_number_list_body(&body)
}

fn yaml_int_vector(yaml: &str, key: &str) -> Vec<usize> {
    yaml_float_vector(yaml, key)
        .into_iter()
        .map(|value| value as usize)
        .collect()
}

fn mat_vec_mul(y: &[f64], rows: usize, cols: usize, x: &[f64]) -> Vec<f64> {
    let mut out = vec![0.0; rows];
    for r in 0..rows {
        let mut sum = 0.0;
        for c in 0..cols {
            sum += y[r * cols + c] * x[c];
        }
        out[r] = sum;
    }
    out
}

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().skip(1).collect();
    if has_flag(&args, "--help") || has_flag(&args, "-h") {
        println!(
            "Usage: cargo run --bin 14_verify_identification -- --data id_verify.csv --params identified.yaml [--urdf robot.urdf]\n\
             or: cargo run --bin 14_verify_identification -- --data id_verify.csv --beta beta.txt [--urdf robot.urdf] [--no-friction]"
        );
        return Ok(());
    }

    let data_path = arg_value(&args, "--data").ok_or("--data is required")?;
    let params_path = arg_value(&args, "--params");
    let beta_path = arg_value(&args, "--beta");
    if params_path.is_none() && beta_path.is_none() {
        return Err("either --params or --beta is required".into());
    }

    let mut mode = "full".to_string();
    let mut include_friction = !has_flag(&args, "--no-friction");
    let mut coulomb_eps = parse_float_arg(&args, "--coulomb-eps", 1e-3);
    let mut selected_columns: Vec<usize> = Vec::new();
    let mut beta: Vec<f64>;
    let urdf_path: PathBuf;

    if let Some(params_path) = params_path {
        let yaml = fs::read_to_string(params_path)?;
        mode = yaml_string(&yaml, "mode", "full");
        include_friction = yaml_bool(&yaml, "include_friction", true);
        if has_flag(&args, "--no-friction") {
            include_friction = false;
        }
        if let Some(value) = yaml_value(&yaml, "coulomb_eps") {
            coulomb_eps = value.parse::<f64>().unwrap_or(coulomb_eps);
        }
        urdf_path = arg_value(&args, "--urdf")
            .map(PathBuf::from)
            .unwrap_or_else(|| PathBuf::from(yaml_string(&yaml, "input_urdf", "")));
        if mode == "base" {
            beta = yaml_float_vector(&yaml, "beta_base");
            if beta.is_empty() {
                beta = yaml_float_vector(&yaml, "beta");
            }
            selected_columns = yaml_int_vector(&yaml, "selected_columns");
        } else if mode == "full" {
            beta = yaml_float_vector(&yaml, "beta");
        } else {
            return Err("--params must contain mode full or base".into());
        }
    } else {
        let text = fs::read_to_string(beta_path.unwrap())?;
        beta = text
            .split_whitespace()
            .map(|value| value.parse::<f64>())
            .collect::<Result<_, _>>()?;
        urdf_path = parse_urdf_path(&args);
    }
    if beta.is_empty() {
        return Err("no beta values found".into());
    }

    let model = MathModel::load(&urdf_path)?;
    let dataset = load_csv(&data_path, model.nq)?;
    let rows = dataset.samples * model.nq;
    let cols = model.num_total_parameters(include_friction)?;
    let y = model.build_regression_matrix(
        &dataset.q,
        &dataset.dq,
        &dataset.ddq,
        dataset.samples,
        include_friction,
        coulomb_eps,
    )?;
    let tau = model.stack_tau_samples(&dataset.tau, dataset.samples)?;
    let tau_pred = if mode == "base" {
        if selected_columns.len() != beta.len() {
            return Err("base beta length does not match selected_columns".into());
        }
        let mut out = vec![0.0; rows];
        for r in 0..rows {
            let mut sum = 0.0;
            for (i, col) in selected_columns.iter().copied().enumerate() {
                if col >= cols {
                    return Err("selected column out of range".into());
                }
                sum += y[r * cols + col] * beta[i];
            }
            out[r] = sum;
        }
        out
    } else {
        if beta.len() != cols {
            return Err("beta length does not match regressor columns".into());
        }
        mat_vec_mul(&y, rows, cols, &beta)
    };
    let (metrics, per_joint_rmse, per_joint_mae) = model.regression_metrics(&tau, &tau_pred)?;
    println!(
        "verify samples={} rmse={:.6} mae={:.6} max_abs={:.6} r2={:.6}",
        dataset.samples, metrics.rmse, metrics.mae, metrics.max_abs, metrics.r2
    );
    println!("per-joint rmse: {:?}", per_joint_rmse);
    println!("per-joint mae:  {:?}", per_joint_mae);
    Ok(())
}
